src/parser/parser.rs

```diff
@@ -123,81 +143,98 @@ fn parse_chain_in_operand(rule: Pair<Rule>) -> Operand {
             }
         }
         jp => Operand::Dynamic(Box::new(jp)),
-    }
+    };
+    Ok(parsed_chain)
 }
 
-fn parse_filter_index(pair: Pair<Rule>) -> JsonPathIndex {
-    JsonPathIndex::Filter(parse_logic(pair.into_inner()))
+fn parse_filter_index(pair: Pair<Rule>) -> Result<JsonPathIndex, JsonPathParserError> {
+    Ok(JsonPathIndex::Filter(parse_logic(pair.into_inner())?))
 }
 
-fn parse_logic(mut pairs: Pairs<Rule>) -> FilterExpression {
+fn parse_logic(pairs: Pairs<Rule>) -> Result<FilterExpression, JsonPathParserError> {
     let mut expr: Option<FilterExpression> = None;
-    while pairs.peek().is_some() {
-        let next_expr = parse_logic_and(pairs.next().unwrap().into_inner());
+    let error_message = format!("Failed to parse logical expression: {:?}", pairs);
```

Comment:
`error_message` has to be constructed before the call to `pair.into_inner()` as `into_inner()` takes ownership and moves the value

Reply:
yeah, sure